File: 2022/Sources/advent2022/Day10.swift
struct Day10 : Day {

    func execute (_ input :[String], _ sampler :(Int, Int) -> Void) {
        var reg = 1
        var cycle = 1
        sampler(cycle, reg)
        for line in input {
            cycle += 1
            if line.starts(with: "addx ") {
                sampler(cycle, reg)
                reg += Int(line.dropFirst(5))!
                cycle += 1
            }
            sampler(cycle, reg)   
        }
    }

    func part1 () throws -> String {
        var signal = 0
        execute(try readInput(10), { (cycle, reg) in 
            if (cycle-20) % 40 == 0 { signal += cycle*reg }
        })
        return String(signal)
    }

    func part2 () throws -> String { 
        var lines = [String]()
        execute(try readInput(10), { (cycle, reg) in 
            let line = (cycle-1) / 40
            if line >= lines.count { lines.append(String()) }
            let col = (cycle-1) % 40
            lines[line].append(reg >= col-1 && reg <= col+1 ? "#" : " ")
        })
        return "\n" + lines.joined(separator: "\n")
     }
}

File: 2022/scala/src/main/scala/Day10.scala
object Day10 extends Day(10):

  def execute (input :Seq[String], sampler :(Int, Int) => Unit) =
    var reg = 1 ; var cycle = 1
    sampler(cycle, reg)
    for (line <- input)
      cycle += 1
      if line startsWith "addx " then
        sampler(cycle, reg)
        reg += line.drop(5).toInt
        cycle += 1
      sampler(cycle, reg)

  override def answer1 (input :Seq[String]) =
    var signal = 0
    execute(input, { (cycle, reg) =>
      if (cycle-20) % 40 == 0 then signal += cycle*reg
    })
    signal

  override def answer2 (input :Seq[String]) =
    var line = "" ; var lines = List[String]()
    execute(input, { (cycle, reg) =>
      val col = (cycle-1) % 40
      line += (if reg >= col-1 && reg <= col+1 then "#" else " ")
      if col == 39 then
        lines = line :: lines
        line = ""
    })
    "\n" + lines.reverse.mkString("\n")
